datasette/static/datasette-manager.js

```diff
@@ -0,0 +1,202 @@
+  aboveTablePanel: '.above-table-panel',
+
+  // These could have multiple matches
+  /** Used for selecting table headers. Use getColumnActions if you want to add menu items. */
+  tableHeaders: `table.rows-and-columns th`,
+
+  /** Used to add "where"  clauses to query using direct manipulation */
+  filterRows: ".filter-row",
+  /** Used to show top available enum values for a column ("facets") */
+  facetResults: ".facet-results [data-column]",
+};
+
+
+/**
+ * Monolith class for interacting with Datasette JS API
+ * Imported with DEFER, runs after main document parsed
+ */
+const datasetteManager = {
+  VERSION: 'TODO_INJECT_VERSION_OR_ENDPOINT_FROM_SERVER_OR_AT_BUILD_TIME',
+
+
+  // TODO: Should order of registration matter?
+
+  // Should plugins be allowed to clobber others or is it last-in takes priority?
+  // Does pluginMetadata need to be serializable, or can we let it be stateful / have functions?
+  plugins: new Map(),
+
+  registerPlugin: (name, pluginMetadata) => {
+    if (datasetteManager.plugins.get(name)) {
+      console.warn(`Warning -> plugin ${name} was redefined`);
+    }
+    datasetteManager.plugins.set(name, pluginMetadata);
+
+    // If the plugin partipates in the panel... update the panel.
+    if (pluginMetadata.getAboveTablePanelConfigs) {
+      datasetteManager.renderAboveTablePanel();
+    }
+  },
+
+  /**
+   * New DOM elements created each time the button is clicked so the data is not stale.
+   * Items
+   *  - must provide label (text)
+   *  - might provide href (string) or an onclick ((evt) => void)
+   *
+   * columnMeta is metadata stored on the column header (TH) as a DOMStringMap
+   * - column: string
+   * - columnNotNull: 0 or 1
+   * - columnType: sqlite datatype enum (text, number, etc)
+   * - isPk: 0 or 1
+   */
+  getColumnActions: (columnMeta) => {
+    let items = [];
+    // Accept function that returns list of items with keys
+    // Required: label (text)
+    // Optional: onClick or href
+    datasetteManager.plugins.forEach(plugin => {
+      if (plugin.getColumnActions) {
+        // Plugins can provide multiple items if they want
+        // If multiple try to create entry with same label, the last one deletes the others
+        items.push(...plugin.getColumnActions(columnMeta));
+      }
+    });
+    // TODO: Validate item configs and give informative error message if missing keys.
+    return items;
+  },
+
+  /**
+   * In MVP, each plugin can only have 1 instance.
+   * In future, panels could be repeated. We omit that for now since so many plugins depend on
+   * shared URL state, so having multiple instances of plugin at same time is problematic.
+   * Currently, we never destroy any panels, we just hide them.
+   *
+   * TODO: nicer panel css, show panel selection state.
+   * TODO: does this hook need to take any arguments?
+   */
+  renderAboveTablePanel: () => {
+
+    const aboveTablePanel = document.querySelector(DOM_SELECTORS.aboveTablePanel);
```

Comment:
Interesting- in those cases, do you think it still makes sense for the plugin to attempt to display (in which case there should be a fallback mount point, or the mount point should always show up regardless of whether there is data), or should a panel like this only show up for views that have some rows?

I think it's the first case, but wanted to check what you think.

Reply:
Decided to always provide a hook so this won't be an issue

https://github.com/simonw/datasette/pull/2052/commits/06b4829035d5fc48b896d4afa5efdf750cf6d895

but also added a bit of JS safety in case the template structure changes in the future.

https://github.com/simonw/datasette/pull/2052/commits/cf504fe49207ec71e0f66340e86fc8d697b437ab